diff --git a/build-aux/config.sub b/build-aux/config.sub
index d74fb6d..a4f2a86 100755
--- a/build-aux/config.sub
+++ b/build-aux/config.sub
@@ -1719,6 +1719,7 @@ case $os in
 	# The portable systems comes first.
 	# Each alternative MUST end in a * to match a version number.
 	gnu* | android* | bsd* | mach* | minix* | genix* | ultrix* | irix* \
+	     | duckos* \
 	     | *vms* | esix* | aix* | cnk* | sunos | sunos[34]* \
 	     | hpux* | unos* | osf* | luna* | dgux* | auroraux* | solaris* \
 	     | sym* |  plan9* | psp* | sim* | xray* | os68k* | v88r* \
